clash-lib/src/Clash/Primitives/Prelude.hs

```diff
@@ -0,0 +1,56 @@
+{-|
+  Copyright   :  (C) 2022     , Myrtle.ai,
+                     2023     , QBayLogic B.V.,
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox functions for primitives in one of the @Prelude@ modules.
+-}
+
+{-# LANGUAGE TemplateHaskellQuotes #-}
+
+module Clash.Primitives.Prelude
+  ( clashCompileErrorBBF
+  ) where
+
+import Control.Monad.State (State)
+import GHC.Stack (HasCallStack)
+import Prelude
+
+import Clash.Backend (Backend)
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.BlackBox.Util (exprToString)
+import Clash.Netlist.Types
+  (BlackBox(..), BlackBoxContext(..), TemplateFunction(..))
+import Data.Text.Prettyprint.Doc.Extra (Doc)
+
+clashCompileErrorBBF :: HasCallStack => BlackBoxFunction
+clashCompileErrorBBF _isD _primName _args _ty =
+  pure (Right (emptyBlackBoxMeta, bb))
+ where
+  bb = BBFunction (show 'clashCompileErrorBBF) 0 clashCompileErrorTF
+  clashCompileErrorTF
+    | ( _hasCallStack
+      : msg
+      : _ ) <- [0..]
+    = TemplateFunction [msg] (const True) clashCompileErrorTemplate
+    | otherwise = undefined
+
+clashCompileErrorTemplate ::
+  HasCallStack =>
+  Backend s =>
+  BlackBoxContext ->
+  State s Doc
+clashCompileErrorTemplate bbCtx
+  | [ _hasCallStack
+    , (msgE, _, _)
+    ] <- bbInputs bbCtx
+  = case exprToString msgE of
```

Comment:
You could reuse `TermLiteral` in `clashCompileErrorBBF`; this would save you from having to write the error message yourself.